pkg/osquery/tables/ioreg/ioreg.go

```diff
@@ -0,0 +1,193 @@
+//+build darwin
+
+// Package ioreg provides a tablle wrapper around the `ioreg` macOS
+// command.
+//
+// As the returned data is a complex nested plist, this uses the
+// dataflatten tooling. (See
+// https://godoc.org/github.com/kolide/launcher/pkg/dataflatten)
+
+package ioreg
+
+import (
+	"bytes"
+	"context"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/kolide/launcher/pkg/dataflatten"
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/kolide/osquery-go"
+	"github.com/kolide/osquery-go/plugin/table"
+	"github.com/pkg/errors"
+)
+
+const ioregPath = "/usr/sbin/ioreg"
+
+const allowedCharacters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
+
+type Table struct {
+	client    *osquery.ExtensionManagerClient
+	logger    log.Logger
+	tableName string
+}
+
+func TablePlugin(client *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+
+	columns := []table.ColumnDefinition{
+		table.TextColumn("fullkey"),
+		table.TextColumn("parent"),
+		table.TextColumn("key"),
+		table.TextColumn("value"),
+		table.TextColumn("query"),
+
+		// ioreg options. It feels awkward using the single
+		// character form, but the full names are too generic
+		// and conflict with our established dataflatten columns
+		table.TextColumn("c"),
+		table.IntegerColumn("d"),
+		table.TextColumn("k"),
+		table.TextColumn("n"),
+		table.TextColumn("p"),
+		table.IntegerColumn("r"), // boolean
```

Comment:
What do you think about something like: 
```suggestion
		table.TextColumn("ioreg_class"),
		table.IntegerColumn("ioreg_depth"),
		table.TextColumn("ioreg_key"),
		table.TextColumn("ioreg_name"),
		table.TextColumn("ioreg_plane"),
		table.IntegerColumn("ioreg_root"), // boolean
```

it's a bit wordy perhaps, but the single char names are really opaque..

we could also do something like: 
```go
		table.TextColumn("class_option"),
		table.IntegerColumn("depth_option"),
		table.TextColumn("key_option"),
		table.TextColumn("name_option"),
		table.TextColumn("plane_option"),
		table.IntegerColumn("root_option"), // boolean
```

Reply:
I thought about it. I didn't really like any option, so I went with succinct. On the sql side, it ends up feeling pretty similar to the command line. `select * from ioreg where  c = 'IOHIDSystem and d = 4`

If you or @FritzX6 feels strongly, I don't mind changing it. But today I'm in a succinct mood. 

Either way. I think the giant set of nested loops needs a better idiom, but I don't have one yet. I guess I could take the first value, and log a warning. But I feel like the nesting here is more correct.

Reply:
I'd feel better about the succinct route if we had some documentation I think. 

I'm with you re: loops, that many nested loops is a pretty strong smell, but it seems to produce the correct behavior...

if the constraints were all `AND`-ed together, we could probably just combine them into a single exec of ioreg. If they're `OR`-ed though, we probably need the multiple calls. As I type that out though, the sqlite engine probably takes care of that for us, so we'd still need to get all the possible rows...

Reply:
Probably also worth noting these aren't output columns -- they're input ones, ideally they'd be hidden as well, but `osquery-go` doesn't support that yet.

Reply:
Semantically, they should be `OR`. I think it's probably wrong to specify more than one, but I don't like breaking the sql assumption of being able to say things like `foo in (a, b)`

What kind of documentation? For better or worse, we don't really document any of these tables.

Reply:
seeing as they're input columns, naming matters a lot less to me. I agree that hiding them would be nice, and as mentioned elsewhere, using this table is going to require `man ioreg` - that's fine. 

If/when we upstream this we can revisit documenting them more.

Reply:
Made the comment more explicit.